src/aabb.rs

```diff
@@ -0,0 +1,175 @@
+//! Axis Aligned Bounding Boxes.
+
+use nalgebra::{Point3, Vector3};
+use std::f32;
+use std::ops::Index;
+
+/// AABB struct.
+#[derive(Debug, Copy, Clone)]
+pub struct AABB {
+    // Minimum coordinates
+    pub min: Point3<f32>,
+
+    // Maximum coordinates
+    pub max: Point3<f32>,
+}
+
+/// A trait implemented by things which can be bounded by an `AABB`.
+pub trait Bounded {
+    fn aabb(&self) -> AABB;
+}
+
+impl AABB {
+    /// Creates a new `AABB` with the given bounds.
+    pub fn new(min: Point3<f32>, max: Point3<f32>) -> AABB {
+        AABB {
+            min: min,
+            max: max,
+        }
+    }
+
+    /// Creates a new empty `AABB`.
+    pub fn empty() -> AABB {
+        AABB {
+            min: Point3::new(f32::INFINITY, f32::INFINITY, f32::INFINITY),
+            max: Point3::new(f32::NEG_INFINITY, f32::NEG_INFINITY, f32::NEG_INFINITY),
+        }
+    }
+
+    /// Returns true if the `Point3` is inside the `AABB`.
+    pub fn contains(&self, p: &Point3<f32>) -> bool {
+        p.x >= self.min.x && p.x <= self.max.x && p.y >= self.min.y && p.y <= self.max.y &&
+        p.z >= self.min.z && p.z <= self.max.z
+    }
+
+    /// Returns true if the `Point3` is approximately inside the `AABB`
+    /// with respect to some `epsilon`.
+    pub fn approx_contains_eps(&self, p: &Point3<f32>, epsilon: f32) -> bool {
+        (p.x - self.min.x) > -epsilon && (p.x - self.max.x) < epsilon &&
+        (p.y - self.min.y) > -epsilon && (p.y - self.max.y) < epsilon &&
+        (p.z - self.min.z) > -epsilon && (p.z - self.max.z) < epsilon
+    }
+
+    /// Returns a new minimal `AABB` which contains both this `AABB` and `other`.
+    pub fn union_aabb(&self, other: &AABB) -> AABB {
```

Comment:
I think this should just be called `union` since `union_point` will be called `grow` so there is no chance of confusing this.